### PR TITLE
Issue 130 : Option to manually select target

### PR DESCRIPTION
**Pull request by poojasahu10**

### CODE REVIEW

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- wplay/onlinetracker.py  1
- wplay/__main__.py  1
- wplay/wchat.py  1
- wplay/messageblast.py  1
- wplay/messagetimer.py  1
- wplay/utils/target_select.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333&bid=16826508)

File: wplay/utils/target_select.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    target_focused_title = await __get_focused_target_title(page)
+    await __wait_for_message_area(page)
+    __print_selected_target_title(target_focused_title)
+    return target_focused_title
+# endregion
+
+# region SELECT TARGET   
+
+def __print_manual_selection_info():
+    print(f"You've to manually select the target from whatsapp web")
```
Comment (rpotter12) [inline wplay/utils/target_select.py]: Codacy Issue found: [Possible SQL injection vector through string-based query construction.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333)

File: wplay/utils/target_select.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        print(f'No target selected! Error: {str(e)}')
+        exit()
+    return target_focused_title
+
+async def __wait_for_message_area(page):
+    try:
+        await page.waitForSelector(whatsapp_selectors_dict['message_area'])
+    except Exception as e:
+        print(f"You don't belong this group anymore! Error: {str(e)}")   
+
```
Comment (rpotter12) [inline wplay/utils/target_select.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333)

File: wplay/utils/target_select.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    except Exception as e:
+        print(f'No target selected! Error: {str(e)}')
+        exit()
+    return target_focused_title
+
+async def __wait_for_message_area(page):
+    try:
+        await page.waitForSelector(whatsapp_selectors_dict['message_area'])
+    except Exception as e:
+        print(f"You don't belong this group anymore! Error: {str(e)}")
```
Comment (rpotter12) [inline wplay/utils/target_select.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333)

File: wplay/utils/target_select.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+async def manual_select_target(page, hide_groups=False):
+    __print_manual_selection_info()
+    await __open_new_chat(page)
+    target_focused_title = await __get_focused_target_title(page)
+    await __wait_for_message_area(page)
+    __print_selected_target_title(target_focused_title)
+    return target_focused_title
+# endregion
+
+# region SELECT TARGET
```
Comment (rpotter12) [inline wplay/utils/target_select.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- wplay/utils/target_select.py  2
- wplay/onlinetracker.py  1
- wplay/messagetimer.py  1
- wplay/wchat.py  1
- wplay/messageblast.py  1
- wplay/__main__.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333&bid=16826508)

File: wplay/utils/target_select.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+def __print_selected_target_title(target_focused_title):
+    print(f"You've selected the target named by: {target_focused_title}")
+
+async def __open_new_chat(page):
+    await page.waitForSelector(
+        whatsapp_selectors_dict['new_chat_button'],
+        visible=True,
+        timeout=0
+    )
+
```
Comment (rpotter12) [inline wplay/utils/target_select.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333)

File: wplay/utils/target_select.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    target_focused_title = await __get_focused_target_title(page)
+    await __wait_for_message_area(page)
+    __print_selected_target_title(target_focused_title)
+    return target_focused_title
+# endregion
+
+# region SELECT TARGET
+def __print_manual_selection_info():
+    print(f"You've to go to whatsapp web and select target manually")
+
```
Comment (rpotter12) [inline wplay/utils/target_select.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- wplay/messagetimer.py  1
- wplay/utils/target_select.py  2
- wplay/__main__.py  1
- wplay/messageblast.py  1
- wplay/onlinetracker.py  1
- wplay/wchat.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333&bid=16826508)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wplay/__main__.py  1
- wplay/onlinetracker.py  1
- wplay/messagetimer.py  1
- wplay/utils/target_select.py  2
- wplay/messageblast.py  1
- wplay/wchat.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333&bid=16826508)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wplay/onlinetracker.py  1
- wplay/messageblast.py  1
- wplay/utils/target_select.py  2
- wplay/__main__.py  1
- wplay/wchat.py  1
- wplay/messagetimer.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333&bid=16826508)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wplay/utils/target_select.py  2
- wplay/onlinetracker.py  1
- wplay/messagetimer.py  1
- wplay/__main__.py  1
- wplay/messageblast.py  1
- wplay/wchat.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333&bid=16826508)

**Comment by xandao6** [comment]
Looks good, but let's wait for some selectors to be fixed first. WhatsApp updated the site and changed some selectors.

**Comment by xandao6** [comment]
image
Add timeout=0 in this selector.

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wplay/messagetimer.py  1
- wplay/messageblast.py  1
- wplay/wchat.py  1
- wplay/__main__.py  1
- wplay/onlinetracker.py  1
- wplay/utils/target_select.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333&bid=16826508)

**Comment by poojasahu10** [comment]
@xandao6 : Timeout made to zero. Kindly review.

**Comment by poojasahu10** [comment]
@rpotter12 : Hi, tracking by name might be broken because of selectors change in whatsapp update. Because that option does not work even for pip package of wplay. Kindly confirm or else I shall dig more to changes in this PR.

**Comment by poojasahu10** [comment]
@rpotter12 : Let me rebase with master and try if selectors change is merged. :)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wplay/__main__.py  1
- wplay/messageblast.py  1
- wplay/wchat.py  1
- wplay/messagetimer.py  1
- wplay/utils/target_select.py  2
- wplay/onlinetracker.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5117333&bid=16826508)

**Comment by poojasahu10** [comment]
@rpotter12 : I get target search failure after rebase with master upstream and the same issue exists in the master too. Shouldn't we raise another issue or merge with this one and solve. I have attached the screenshot for the same. Thanks.
SearchTarget

**Comment by rpotter12** [comment]
@poojasahu10 please test your manually target selecting option. It is not working. :)

**Comment by poojasahu10** [comment]
@rpotter12 : I did try to test for edge cases for manually selecting the target but could not find any issues. Can you share the screenshot of the issue? Thanks.

**Comment by rpotter12** [comment]
@poojasahu10 I is working good :)
